File: tests/unit/backoff.test.ts
/**
 * Per-origin rate-limit backoff (0.3.0). The motivating case is a corpus of ~72 GitHub
 * repos ingested `concurrency` at a time: every one of them talks to `api.github.com`, so
 * the retry policy has to be a property of the HOST, not of one client or one repo.
 *
 * Nothing here sleeps: the clock and the sleep are injected, and the assertions are about
 * what was scheduled and how many requests actually went out.
 */
import { describe, expect, it } from 'vitest';
import { OriginBackoff, originOf, type BackoffOptions } from '../../src/lib/importers/backoff';
import { JsonClient } from '../../src/lib/importers/http';
import { ImporterError } from '../../src/lib/importers/index';

/** A backoff with a fake clock: sleeping advances time instead of waiting. */
function fakeBackoff(overrides: BackoffOptions = {}) {
  let now = 1_000_000;
  const slept: number[] = [];
  const backoff = new OriginBackoff({
    baseDelayMs: 1_000,
    maxDelayMs: 60_000,
    maxAttempts: 4,
    jitter: () => 0, // deterministic: assert the exact exponential ladder
    now: () => now,
    sleepImpl: async (ms) => {
      slept.push(ms);
      now += ms;
    },
    ...overrides,
  });
  return { backoff, slept, advance: (ms: number) => (now += ms), at: () => now };
}

/** A fetchImpl that replays a scripted list of responses and counts the calls. */
function scriptedFetch(script: Array<{ status: number; headers?: Record<string, string>; body?: string }>) {
  const urls: string[] = [];
  const fetchImpl = (async (url: string) => {
    urls.push(String(url));
    const step = script[Math.min(urls.length - 1, script.length - 1)]!;
    return new Response(step.body ?? '{"ok":true}', {
      status: step.status,
      headers: step.headers ?? {},
    });
  }) as unknown as typeof fetch;
  return { fetchImpl, urls };
}

const RATE_LIMIT_HEADERS = { 'x-ratelimit-remaining': '0' };

describe('OriginBackoff', () => {
  it('grows exponentially per consecutive rate limit on the same origin', async () => {
    const { backoff, slept } = fakeBackoff();
    // No Retry-After: the delay comes from the consecutive-failure count. jitter() is 0, so
    // the ladder is exactly base, base*2, base*4.
    expect(await backoff.noteRateLimit('https://api.github.com', undefined, 1)).toBe(true);
    expect(await backoff.noteRateLimit('https://api.github.com', undefined, 2)).toBe(true);
    expect(await backoff.noteRateLimit('https://api.github.com', undefined, 3)).toBe(true);
    expect(slept).toEqual([1_000, 2_000, 4_000]);
  });

  it('honours the provider’s Retry-After instead of guessing', async () => {
    const { backoff, slept } = fakeBackoff();
    expect(await backoff.noteRateLimit('https://api.github.com', 7, 1)).toBe(true);
    expect(slept).toEqual([7_000]);
  });

  it('stops retrying once the attempt budget is spent', async () => {
    const { backoff } = fakeBackoff({ maxAttempts: 2 });
    expect(await backoff.noteRateLimit('https://api.github.com', 1, 1)).toBe(true);
    expect(await backoff.noteRateLimit('https://api.github.com', 1, 2)).toBe(false);
  });

  it('blocks (rather than sleeps through) a limit longer than the ceiling', async () => {
    // GitHub's hourly quota says "come back in 40 minutes". Sleeping that out would hang the
    // build; retrying would waste every repo's attempts. The origin is marked blocked.
    const { backoff, slept } = fakeBackoff();
    expect(await backoff.noteRateLimit('https://api.github.com', 2400, 1)).toBe(false);
    expect(slept).toEqual([]); // nothing was waited on
    expect(backoff.isBlocked('https://api.github.com')).toBe(true);

    await expect(backoff.beforeRequest('https://api.github.com', 'GET /repos/x')).rejects.toThrow(
      /rate-limited for another/,
    );
    // ...and it is the kind that maps onto `remote-rate-limited` + the cached fallback
    await backoff
      .beforeRequest('https://api.github.com', 'GET /repos/x')
      .catch((e: unknown) => expect((e as ImporterError).kind).toBe('rate-limit'));
  });

  it('isolates origins: one forge backing off never delays another', async () => {
    const { backoff, slept } = fakeBackoff();
    await backoff.noteRateLimit('https://api.github.com', 30, 1);
    expect(slept).toEqual([30_000]);

    // codeberg has its own (clean) state — no wait at all
    await backoff.beforeRequest('https://codeberg.org', 'GET /x');
    expect(slept).toEqual([30_000]);
    expect(backoff.peek('https://codeberg.org').failures).toBe(0);
  });

  it('a success clears the consecutive-failure count', async () => {
    const { backoff, slept } = fakeBackoff();
    await backoff.noteRateLimit('https://api.github.com', undefined, 1);
    await backoff.noteRateLimit('https://api.github.com', undefined, 2);
    backoff.noteSuccess('https://api.github.com');
    slept.length = 0;
    await backoff.noteRateLimit('https://api.github.com', undefined, 1);
    expect(slept).toEqual([1_000]); // back to the first rung, not the third
  });

  it('holds one shared window per origin, so every caller waits behind the same timer', async () => {
    // Modelling real concurrency: caller A hits the limit and is still sleeping (its sleep
    // never resolves here) while caller B arrives. B must see A's window, not a fresh one.
    let now = 1_000_000;
    const backoff = new OriginBackoff({
      baseDelayMs: 1_000,
      maxDelayMs: 60_000,
      jitter: () => 0,
      now: () => now,
      sleepImpl: () => new Promise<void>(() => {}), // A stays parked in its backoff
    });

    void backoff.noteRateLimit('https://api.github.com', 30, 1);
    await Promise.resolve();

    // The window is open and is a property of the ORIGIN — B reads the same remaining time.
    expect(backoff.peek('https://api.github.com').waitingMs).toBe(30_000);
    now += 10_000;
    expect(backoff.peek('https://api.github.com').waitingMs).toBe(20_000);
    // ...and a different origin is untouched by it
    expect(backoff.peek('https://codeberg.org').waitingMs).toBe(0);

    now += 20_000;
    expect(backoff.peek('https://api.github.com').waitingMs).toBe(0);
  });

  it('originOf keys on scheme+host+port, and survives a non-URL', () => {
    expect(originOf('https://api.github.com/repos/a/b')).toBe('https://api.github.com');
    expect(originOf('https://git.example.com:3000/api/v1/x')).toBe('https://git.example.com:3000');
    expect(originOf('https://gitlab.com/api/v4/y')).toBe('https://gitlab.com');
    // different hosts must not collide
    expect(originOf('https://codeberg.org/api/v1/x')).not.toBe(originOf('https://api.github.com/x'));
    expect(originOf('not a url')).toBe('not a url');
  });
});

describe('JsonClient rate-limit handling', () => {
  const client = (fetchImpl: typeof fetch, backoff: OriginBackoff) =>
    new JsonClient({ auth: 'bearer', fetchImpl, retryDelayMs: 0, backoff });

  it('retries a 429 and succeeds, without spending the 5xx retry budget', async () => {
    const { backoff } = fakeBackoff();
    const { fetchImpl, urls } = scriptedFetch([
      { status: 429, headers: { 'retry-after': '2' } },
      { status: 429, headers: { 'retry-after': '2' } },
      { status: 200, body: '{"name":"ok"}' },
    ]);
    const result = await client(fetchImpl, backoff).get<{ name: string }>('https://api.github.com/repos/a/b');
    expect(result.name).toBe('ok');
    expect(urls.length).toBe(3); // two rate-limited attempts then the good one
  });

  it('treats GitHub’s 403-with-no-quota-left as a rate limit, not an auth failure', async () => {
    const { backoff } = fakeBackoff();
    const { fetchImpl, urls } = scriptedFetch([
      { status: 403, headers: RATE_LIMIT_HEADERS },
      { status: 200, body: '{"name":"ok"}' },
    ]);
    const result = await client(fetchImpl, backoff).get<{ name: string }>('https://api.github.com/repos/a/b');
    expect(result.name).toBe('ok');
    expect(urls.length).toBe(2);
  });

  it('gives up after the attempt budget and reports a rate limit', async () => {
    const { backoff } = fakeBackoff({ maxAttempts: 2 });
    const { fetchImpl, urls } = scriptedFetch([{ status: 429, headers: { 'retry-after': '1' } }]);
    await expect(
      client(fetchImpl, backoff).get('https://api.github.com/repos/a/b'),
    ).rejects.toMatchObject({ kind: 'rate-limit' });
    expect(urls.length).toBe(2);
  });

  it('once an origin is hard-blocked, later calls cost no requests at all', async () => {
    // The 72-repo case: repo #1 discovers the hourly limit, repos #2..72 must not each make
    // four more doomed calls — they fail fast and fall back to their cached metadata.
    const { backoff } = fakeBackoff();
    const { fetchImpl, urls } = scriptedFetch([{ status: 429, headers: { 'retry-after': '3600' } }]);
    const c = client(fetchImpl, backoff);

    await expect(c.get('https://api.github.com/repos/a/b')).rejects.toMatchObject({ kind: 'rate-limit' });
    const afterFirst = urls.length;

    await expect(c.get('https://api.github.com/repos/c/d')).rejects.toMatchObject({ kind: 'rate-limit' });
    expect(urls.length).toBe(afterFirst); // no new request went out

    // a different forge still works
    const { fetchImpl: otherFetch, urls: otherUrls } = scriptedFetch([{ status: 200, body: '{"name":"ok"}' }]);
    await client(otherFetch, backoff).get('https://codeberg.org/api/v1/repos/a/b');
    expect(otherUrls.length).toBe(1);
  });

  it('does not treat a plain 401 as a rate limit', async () => {
    const { backoff } = fakeBackoff();
    const { fetchImpl, urls } = scriptedFetch([{ status: 401, body: '{"message":"Bad credentials"}' }]);
    await expect(client(fetchImpl, backoff).get('https://api.github.com/repos/a/b')).rejects.toMatchObject({
      kind: 'auth',
    });
    expect(urls.length).toBe(1); // no retry ladder for a bad token
  });
});

File: internal/ingest/backoff_test.go
package ingest

import (
	"context"
	"errors"
	"strings"
	"sync"
	"testing"
	"time"
)

// Per-origin rate-limit backoff. The motivating case is a corpus of ~72 GitHub repos ingested
// `concurrency` at a time: every one of them talks to api.github.com, so the retry policy has to
// be a property of the HOST, not of one client or one repo.
//
// Nothing here sleeps: the clock and the sleep are injected, and the assertions are about what
// was scheduled and how many requests actually went out.

// fakeClock is a backoff whose sleep advances a counter instead of waiting.
type fakeClock struct {
	mu    sync.Mutex
	now   time.Time
	slept []time.Duration
	// park makes sleep never return, modelling a caller still waiting inside its backoff.
	park bool
}

func newFakeBackoff(t testing.TB, opts BackoffOptions) (*OriginBackoff, *fakeClock) {
	t.Helper()
	clock := &fakeClock{now: time.Unix(1_000_000, 0)}
	if opts.BaseDelay == 0 {
		opts.BaseDelay = time.Second
	}
	if opts.MaxDelay == 0 {
		opts.MaxDelay = 60 * time.Second
	}
	if opts.MaxAttempts == 0 {
		opts.MaxAttempts = 4
	}
	// Deterministic: assert the exact exponential ladder.
	opts.Jitter = func() float64 { return 0 }
	opts.Now = clock.read
	opts.Sleep = clock.sleep
	return NewOriginBackoff(opts), clock
}

func (c *fakeClock) read() time.Time {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.now
}

func (c *fakeClock) sleep(ctx context.Context, d time.Duration) error {
	c.mu.Lock()
	c.slept = append(c.slept, d)
	park := c.park
	if !park {
		c.now = c.now.Add(d)
	}
	c.mu.Unlock()
	if park {
		<-ctx.Done()
		return ctx.Err()
	}
	return nil
}

func (c *fakeClock) advance(d time.Duration) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.now = c.now.Add(d)
}

func (c *fakeClock) sleeps() []time.Duration {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make([]time.Duration, len(c.slept))
	copy(out, c.slept)
	return out
}

func wantSleeps(t testing.TB, got []time.Duration, want ...time.Duration) {
	t.Helper()
	if len(got) != len(want) {
		t.Fatalf("slept %v, want %v", got, want)
	}
	for i := range got {
		if got[i] != want[i] {
			t.Fatalf("slept %v, want %v", got, want)
		}
	}
}

const gh = "https://api.github.com"

func TestBackoffGrowsExponentiallyPerConsecutiveRateLimit(t *testing.T) {
	b, clock := newFakeBackoff(t, BackoffOptions{})
	ctx := context.Background()
	// No Retry-After: the delay comes from the consecutive-failure count. Jitter is 0, so the
	// ladder is exactly base, base*2, base*4.
	for attempt := 1; attempt <= 3; attempt++ {
		retry, err := b.NoteRateLimit(ctx, gh, 0, false, attempt)
		if err != nil || !retry {
			t.Fatalf("attempt %d: retry=%v err=%v", attempt, retry, err)
		}
	}
	wantSleeps(t, clock.sleeps(), time.Second, 2*time.Second, 4*time.Second)
}

func TestBackoffHonoursRetryAfter(t *testing.T) {
	b, clock := newFakeBackoff(t, BackoffOptions{})
	retry, err := b.NoteRateLimit(context.Background(), gh, 7, true, 1)
	if err != nil || !retry {
		t.Fatalf("retry=%v err=%v", retry, err)
	}
	wantSleeps(t, clock.sleeps(), 7*time.Second)
}

func TestBackoffStopsAtTheAttemptBudget(t *testing.T) {
	b, _ := newFakeBackoff(t, BackoffOptions{MaxAttempts: 2})
	ctx := context.Background()
	if retry, _ := b.NoteRateLimit(ctx, gh, 1, true, 1); !retry {
		t.Fatal("first rate limit should have been retried")
	}
	if retry, _ := b.NoteRateLimit(ctx, gh, 1, true, 2); retry {
		t.Fatal("the budget was spent; the caller must fall back to its cache")
	}
}

func TestBackoffBlocksRatherThanSleepsThroughALongLimit(t *testing.T) {
	// GitHub's hourly quota says "come back in 40 minutes". Sleeping that out would hang the
	// build; retrying would waste every repo's attempts. The origin is marked blocked instead.
	b, clock := newFakeBackoff(t, BackoffOptions{})
	ctx := context.Background()
	if retry, err := b.NoteRateLimit(ctx, gh, 2400, true, 1); retry || err != nil {
		t.Fatalf("retry=%v err=%v", retry, err)
	}
	wantSleeps(t, clock.sleeps()) // nothing was waited on
	if !b.IsBlocked(gh) {
		t.Fatal("the origin should be blocked")
	}

	err := b.BeforeRequest(ctx, gh, "GET /repos/x")
	if err == nil || !strings.Contains(err.Error(), "rate-limited for another") {
		t.Fatalf("BeforeRequest err = %v", err)
	}
	// ...and it is the kind that maps onto remote-rate-limited plus the cached fallback.
	var ie *ImporterError
	if !errors.As(err, &ie) || ie.Kind != KindRateLimit {
		t.Fatalf("want a rate-limit ImporterError, got %#v", err)
	}
	if ie.RetryAfter == nil || *ie.RetryAfter != 2400 {
		t.Fatalf("RetryAfter = %v", ie.RetryAfter)
	}
}

func TestBackoffIsolatesOrigins(t *testing.T) {
	b, clock := newFakeBackoff(t, BackoffOptions{})
	ctx := context.Background()
	if _, err := b.NoteRateLimit(ctx, gh, 30, true, 1); err != nil {
		t.Fatal(err)
	}
	wantSleeps(t, clock.sleeps(), 30*time.Second)

	// codeberg has its own (clean) state — no wait at all.
	if err := b.BeforeRequest(ctx, "https://codeberg.org", "GET /x"); err != nil {
		t.Fatal(err)
	}
	wantSleeps(t, clock.sleeps(), 30*time.Second)
	if got := b.Peek("https://codeberg.org").Failures; got != 0 {
		t.Fatalf("codeberg failures = %d", got)
	}
}

func TestBackoffSuccessClearsTheFailureCount(t *testing.T) {
	b, clock := newFakeBackoff(t, BackoffOptions{})
	ctx := context.Background()
	_, _ = b.NoteRateLimit(ctx, gh, 0, false, 1)
	_, _ = b.NoteRateLimit(ctx, gh, 0, false, 2)
	b.NoteSuccess(gh)

	before := len(clock.sleeps())
	if _, err := b.NoteRateLimit(ctx, gh, 0, false, 1); err != nil {
		t.Fatal(err)
	}
	got := clock.sleeps()[before:]
	// Back to the first rung, not the third.
	wantSleeps(t, got, time.Second)
}

func TestBackoffHoldsOneSharedWindowPerOrigin(t *testing.T) {
	// Modelling real concurrency: caller A hits the limit and is still sleeping while caller B
	// arrives. B must see A's window, not a fresh one.
	b, clock := newFakeBackoff(t, BackoffOptions{})
	clock.mu.Lock()
	clock.park = true
	clock.mu.Unlock()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	parked := make(chan struct{})
	go func() {
		defer close(parked)
		_, _ = b.NoteRateLimit(ctx, gh, 30, true, 1)
	}()
	// Wait for A to have opened the window before reading it.
	waitFor(t, func() bool { return b.Peek(gh).Waiting > 0 })

	if got := b.Peek(gh).Waiting; got != 30*time.Second {
		t.Fatalf("waiting = %v", got)
	}
	clock.advance(10 * time.Second)
	if got := b.Peek(gh).Waiting; got != 20*time.Second {
		t.Fatalf("waiting after 10s = %v", got)
	}
	// ...and a different origin is untouched by it.
	if got := b.Peek("https://codeberg.org").Waiting; got != 0 {
		t.Fatalf("codeberg waiting = %v", got)
	}
	clock.advance(20 * time.Second)
	if got := b.Peek(gh).Waiting; got != 0 {
		t.Fatalf("waiting after the window = %v", got)
	}
	cancel()
	<-parked
}

func TestOriginOfKeysOnSchemeHostAndPort(t *testing.T) {
	cases := map[string]string{
		"https://api.github.com/repos/a/b":      "https://api.github.com",
		"https://git.example.com:3000/api/v1/x": "https://git.example.com:3000",
		"https://gitlab.com/api/v4/y":           "https://gitlab.com",
		"not a url":                             "not a url",
		"https://codeberg.org/api/v1/repos/a/b": "https://codeberg.org",
	}
	for in, want := range cases {
		if got := OriginOf(in); got != want {
			t.Errorf("OriginOf(%q) = %q, want %q", in, got, want)
		}
	}
	// Different hosts must not collide.
	if OriginOf("https://codeberg.org/api/v1/x") == OriginOf("https://api.github.com/x") {
		t.Error("two forges collapsed onto one origin")
	}
}

/* ---- JSONClient rate-limit handling -------------------------------------- */

func rateLimitClient(doer Doer, backoff *OriginBackoff) *JSONClient {
	return NewJSONClient(JSONClientOptions{
		Auth: AuthBearer, HTTP: doer, RetryDelay: -1, Backoff: backoff,
	})
}

func TestClientRetriesA429WithoutSpendingThe5xxBudget(t *testing.T) {
	b, _ := newFakeBackoff(t, BackoffOptions{})
	doer := newScriptedDoer(
		httpFixtureRoute{Status: 429, Headers: map[string]string{"retry-after": "2"}},
		httpFixtureRoute{Status: 429, Headers: map[string]string{"retry-after": "2"}},
		httpFixtureRoute{Status: 200, Body: `{"name":"ok"}`},
	)
	raw, err := rateLimitClient(doer, b).Get(context.Background(), gh+"/repos/a/b")
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if name := NullIfEmpty(asObject(raw)["name"]); name == nil || *name != "ok" {
		t.Fatalf("name = %v", name)
	}
	if doer.count() != 3 { // two rate-limited attempts then the good one
		t.Fatalf("requests = %d, want 3", doer.count())
	}
}

func TestClientTreatsGithub403WithNoQuotaLeftAsARateLimit(t *testing.T) {
	b, _ := newFakeBackoff(t, BackoffOptions{})
	doer := newScriptedDoer(
		httpFixtureRoute{Status: 403, Headers: map[string]string{"x-ratelimit-remaining": "0"}},
		httpFixtureRoute{Status: 200, Body: `{"name":"ok"}`},
	)
	if _, err := rateLimitClient(doer, b).Get(context.Background(), gh+"/repos/a/b"); err != nil {
		t.Fatalf("Get: %v", err)
	}
	if doer.count() != 2 {
		t.Fatalf("requests = %d, want 2", doer.count())
	}
}

func TestClientGivesUpAfterTheAttemptBudget(t *testing.T) {
	b, _ := newFakeBackoff(t, BackoffOptions{MaxAttempts: 2})
	doer := newScriptedDoer(httpFixtureRoute{Status: 429, Headers: map[string]string{"retry-after": "1"}})
	_, err := rateLimitClient(doer, b).Get(context.Background(), gh+"/repos/a/b")
	assertImporterKind(t, err, KindRateLimit)
	if doer.count() != 2 {
		t.Fatalf("requests = %d, want 2", doer.count())
	}
}

func TestClientHardBlockCostsNoFurtherRequests(t *testing.T) {
	// The 72-repo case: repo #1 discovers the hourly limit, repos #2..72 must not each make four
	// more doomed calls — they fail fast and fall back to their cached metadata.
	b, _ := newFakeBackoff(t, BackoffOptions{})
	doer := newScriptedDoer(httpFixtureRoute{Status: 429, Headers: map[string]string{"retry-after": "3600"}})
	client := rateLimitClient(doer, b)

	_, err := client.Get(context.Background(), gh+"/repos/a/b")
	assertImporterKind(t, err, KindRateLimit)
	afterFirst := doer.count()

	_, err = client.Get(context.Background(), gh+"/repos/c/d")
	assertImporterKind(t, err, KindRateLimit)
	if doer.count() != afterFirst {
		t.Fatalf("a blocked origin still sent %d requests", doer.count()-afterFirst)
	}

	// A different forge still works.
	other := newScriptedDoer(httpFixtureRoute{Status: 200, Body: `{"name":"ok"}`})
	if _, err := rateLimitClient(other, b).Get(context.Background(), "https://codeberg.org/api/v1/repos/a/b"); err != nil {
		t.Fatalf("codeberg: %v", err)
	}
	if other.count() != 1 {
		t.Fatalf("codeberg requests = %d", other.count())
	}
}

func TestClientDoesNotTreatAPlain401AsARateLimit(t *testing.T) {
	b, _ := newFakeBackoff(t, BackoffOptions{})
	doer := newScriptedDoer(httpFixtureRoute{Status: 401, Body: `{"message":"Bad credentials"}`})
	_, err := rateLimitClient(doer, b).Get(context.Background(), gh+"/repos/a/b")
	assertImporterKind(t, err, KindAuth)
	if doer.count() != 1 { // no retry ladder for a bad token
		t.Fatalf("requests = %d, want 1", doer.count())
	}
}

func assertImporterKind(t testing.TB, err error, want ImporterErrorKind) {
	t.Helper()
	var ie *ImporterError
	if !errors.As(err, &ie) {
		t.Fatalf("want an *ImporterError, got %v", err)
	}
	if ie.Kind != want {
		t.Fatalf("kind = %q, want %q (message: %s)", ie.Kind, want, ie.Message)
	}
}

// waitFor spins until cond holds, failing the test rather than hanging forever.
func waitFor(t testing.TB, cond func() bool) {
	t.Helper()
	deadline := time.Now().Add(2 * time.Second)
	for !cond() {
		if time.Now().After(deadline) {
			t.Fatal("condition never became true")
		}
		time.Sleep(time.Millisecond)
	}
}
